pkg/runtime/scheduler/scheduler.go

```diff
@@ -36,183 +42,224 @@ type Options struct {
-
-	lock sync.Mutex
+// Scheduler manages the connection to the cluster of schedulers.
+type Scheduler struct {
+	addresses []string
+	security  security.Handler
+	htarget   healthz.Target
 
-	appCh      chan struct{}
-	appRunning bool
+	cluster *cluster.Cluster
+	clients *clients.Clients
 
-	running atomic.Bool
-	wg      sync.WaitGroup
+	lock     sync.RWMutex
+	readyCh  chan struct{}
+	disabled chan struct{}
 }
 
-func New(opts Options) *Manager {
-	return &Manager{
-		namespace: opts.Namespace,
-		appID:     opts.AppID,
-		actors:    opts.Actors,
-		channels:  opts.Channels,
-		clients:   opts.Clients,
-		appCh:     make(chan struct{}),
+func New(opts Options) *Scheduler {
+	return &Scheduler{
+		addresses: opts.Addresses,
+		security:  opts.Security,
+		cluster: cluster.New(cluster.Options{
+			Namespace: opts.Namespace,
+			AppID:     opts.AppID,
+			Actors:    opts.Actors,
+			Channels:  opts.Channels,
+		}),
+		htarget:  opts.Healthz.AddTarget(),
+		readyCh:  make(chan struct{}),
+		disabled: make(chan struct{}),
 	}
 }
 
-// Run starts watching for job triggers from all scheduler clients.
-func (m *Manager) Run(ctx context.Context) error {
-	if !m.running.CompareAndSwap(false, true) {
-		return errors.New("scheduler manager is already running")
-	}
-	defer m.wg.Wait()
-
-	var typeUpdateCh <-chan []string = make(chan []string)
-	var actorTypes []string
-	if table, err := m.actors.Table(ctx); err == nil {
-		typeUpdateCh, actorTypes = table.SubscribeToTypeUpdates(ctx)
+func (s *Scheduler) Run(ctx context.Context) error {
+	if len(s.addresses) == 0 {
+		s.htarget.Ready()
+		close(s.disabled)
+		<-ctx.Done()
+		return nil
 	}
 
 	for {
-		m.lock.Lock()
-		appRunning := m.appRunning
-		appCh := m.appCh
-		m.lock.Unlock()
-
-		if !appRunning && len(actorTypes) == 0 {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case <-appCh:
-			case actorTypes = <-typeUpdateCh:
-			}
+		// TODO: @joshvanl: handle not WatchHosts unimplemented.
+		// TODO: @joshvanl: don't need to reconnect client if stream is still
+		// connected.
+		stream, addresses, err := s.connSchedulerHosts(ctx)
+		if err != nil {
+			return err
 		}
 
-		m.lock.Lock()
-		appCh = m.appCh
-		appRunning = m.appRunning
-		m.lock.Unlock()
+		err = concurrency.NewRunnerManager(
+			func(ctx context.Context) error {
+				return s.connectClients(ctx, addresses)
+			},
+			func(ctx context.Context) error {
+				if stream == nil {
+					<-ctx.Done()
+					return nil
+				}
+
+				var resp *schedulerv1pb.WatchHostsResponse
+				resp, err = stream.Recv()
+				if err != nil {
+					log.Errorf("Failed to receive scheduler hosts: %s", err)
+					//nolint:nilerr
+					return nil
+				}
+				addresses = make([]string, 0, len(resp.GetHosts()))
+				for _, host := range resp.GetHosts() {
+					addresses = append(addresses, host.GetAddress())
+				}
+
+				log.Infof("Received updated scheduler hosts addresses: %v", addresses)
+
+				return nil
+			},
+		).Run(ctx)
```

Comment:
I don't think I understand this block, how do we ensure `s.connectClients` will run after we've received new addresses? Don't we have a potential race condition on accessing `addresses`?

Reply:
Ok, nevermind, I see it now, `s.connectClients` won't return until there's an error on the stream.
Would be nice to trigger the reconnect automatically without reconnecting, but I see you already have a TODO for that.